src/goImplementations.ts

```diff
@@ -0,0 +1,77 @@
+'use strict';
+
+import vscode = require('vscode');
+import cp = require('child_process');
+import path = require('path');
+import { byteOffsetAt, getBinPath, canonicalizeGOPATHPrefix } from './util';
+import { promptForMissingTool } from './goInstallTools';
+import { goKeywords, isPositionInString } from './util';
+import { getGoRuntimePath, resolvePath } from './goPath';
+
+interface GoListOutput {
+	Dir: string;
+	ImportPath: string;
+}
+
+interface GuruImplementsRef {
+	name: string;
+	pos: string;
+	kind: string;
+}
+
+interface GuruImplementsOutput {
+	type: GuruImplementsRef;
+	to: GuruImplementsRef[];
+	from: GuruImplementsRef[];
+}
+
+export class GoImplementationProvider implements vscode.ImplementationProvider {
+	public provideImplementation(document: vscode.TextDocument, position: vscode.Position, token: vscode.CancellationToken): Thenable<vscode.Definition> {
+		return new Promise<vscode.Definition>((resolve, reject) => {
+			if (token.isCancellationRequested) {
+				return resolve(null);
+			}
+
+			let listProcess = cp.execFile(getGoRuntimePath(), ['list', '-e', '-json'], { cwd: vscode.workspace.rootPath }, (err, stdout, stderr) => {
```

Comment:
The `-scope` argument expects a package pattern, which in this implementation I chose to make recursive (e.g. if the current package according to `go list` is `github.com/my/lib`, the `-scope` will be set to `github.com/my/lib/...` since it seems likely you want to find implementations across all packages in your project relative to the base package).

Here are the docs for scope as applies to `implements`:

```
The -scope flag restricts analysis to the specified packages.
        Its value is a comma-separated list of patterns of these forms:
                golang.org/x/tools/cmd/guru     # a single package
                golang.org/x/tools/...          # all packages beneath dir
                ...                             # the entire workspace.
        A pattern preceded by '-' is negative, so the scope
                encoding/...,-encoding/xml
        matches all encoding packages except encoding/xml.
```

Using `-scope=...` seems the most generic but probably the slowest (although the difference could be negligible for typical projects... for Kubernetes, a random analysis I have chosen takes 13s scoped to `k8s.io/kubernetes/...` and 28s scoped to `...`).

I'm not sure what the right assumptions to make are for the typical user. In my case, I always open a package base directory when working with Go projects in VSCode. I'm sure the upcoming multi-root support will further complicate matters here.

Maybe there's some value in supporting a variety of scope strategies so users can either preset or interactively choose a scope for these sorts of analysis operations with guru?

Seems like no matter what we try to infer we're going to be wrong some of the time, but a default of some kind is clearly necessary.